src/Model/BSONIterator.php

```diff
@@ -0,0 +1,117 @@
+<?php
+/*
+ * Copyright 2018 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Model;
+
+use MongoDB\BSON;
+use MongoDB\Exception\UnexpectedValueException;
+use MongoDB\Model\BSONDocument;
+use Iterator;
+
+/**
+ * Iterator for BSON documents.
+ */
+class BSONIterator implements Iterator
+{
+    private $buffer;
+    private $bufferLength;
+    private $current;
+    private $key = 0;
+    private $position = 0;
+
+    const BSON_SIZE = 4;
+
+    /**
+     * Constructor.
+    *
+     * @internal
+     * @param string $data Concatenated, valid, BSON-encoded documents
+     */
+    public function __construct($data)
+    {
+        $this->buffer = $data;
+        $this->bufferLength = strlen($data);
+    }
+
+    /**
+     */
+    public function current()
+    {
+        return $this->current;
+    }
+
+    /**
+     * @see http://php.net/iterator.key
+     * @return mixed
+     */
+    public function key()
+    {
+        return $this->key;
+    }
+
+    /**
+     * @see http://php.net/iterator.next
+     * @return void
+     */
+    public function next()
+    {
+        $this->key++;
+        $this->current = null;
+        $this->getNextBSON();
+    }
+
+    /**
+     * @see http://php.net/iterator.rewind
+     * @return void
+     */
+    public function rewind()
+    {
+        $this->key = 0;
+        $this->position = 0;
+        $this->current = null;
+        $this->getNextBSON();
+    }
+
+    /**
+     * @see http://php.net/iterator.valid
+     * @return boolean
+     */
+    public function valid()
+    {
+        return $this->current !== null;
+    }
+
+    private function getNextBSON()
+    {
+        if ($this->position === $this->bufferLength) {
+            return;
+        }
+
+        if (($this->bufferLength - $this->position) < 5) {
+            throw new UnexpectedValueException(sprintf('Expected at least %d bytes; %d remaining', self::BSON_SIZE, $this->bufferLength - $this->position));
```

Comment:
I think this error message is actually incorrect. We're expecting at least 5 bytes (i.e. smallest possible document), which would include the 32-bit integer denoting the root document size and a trailing null byte.

PHP 5.5 won't allow us to define a constant relative to `self::BSON_SIZE` ([constant expressions](http://php.net/manual/en/language.oop5.constants.php) were introduced in 5.6), but I'd suggest a new `MIN_DOCUMENT_LENGTH` constant defined to 5, which we can reference in this error message and the conditional above (instead of using a [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming))).